Re-implement unread counts

babolivier opened the pull request:
Second attempt at implementing unread counts, this time without relying on push rules. ~This is also backing out the MSC2625 implementation, which we likely won't end up using~ reverting the MSC2625 implem has been moved to #7761.

~~Best reviewed commit by commit.~~

Fixes https://github.com/matrix-org/synapse/issues/7741

richvdh commented:
> This is also backing out the MSC2625 implementation

which, for linking, was #7673.

t3chguy commented on synapse/storage/data_stores/main/event_push_actions.py:
```diff
@@ -465,15 +428,14 @@ def add_push_actions_to_staging(self, event_id, user_id_actions):
 
         # This is a helper function for generating the necessary tuple that
-        # can be used to insert into the `event_push_actions_staging` table.
+        # can be used to inert into the `event_push_actions_staging` table.
```
probably no harm in fixing the typo ;)

richvdh commented on synapse/storage/data_stores/main/events.py:
```diff
@@ -1614,3 +1626,87 @@ def f(txn, stream_ordering):
+        """
+        content = event.content
+
+        is_edit = (
+            content.get("m.relates_to", {}).get("rel_type") == RelationTypes.REPLACE
+        )
+        is_notice = not event.is_state() and content.get("msgtype") == "m.notice"
+
+        # We don't want rejected or soft-failed events, edits or notices to be marked
+        # unread.
+        if (
+            context.rejected
+            or is_edit
+            or is_notice
+            or event.internal_metadata.is_soft_failed()
+        ):
+            return
```
I think this would be clearer (and more efficient) as a set of separate `if` conditions:

```
if context.rejected or event.internal_metadata.is_soft_failed():
    # exclude rejected and soft-failed events
    return

if content.get("m.relates_to", {}).get("rel_type") == RelationTypes.REPLACE:
    # it's an edit: ignore
    return

if not event.is_state() and content.get("msgtype") == "m.notice":
   # etc
```

richvdh commented on synapse/storage/data_stores/main/events.py:
```diff
@@ -61,6 +62,13 @@
     ["type", "origin_type", "origin_entity"],
 )
 
+STATE_EVENT_TYPES_TO_MARK_UNREAD = [
+    EventTypes.PowerLevels,
+    EventTypes.Topic,
+    EventTypes.Name,
+    EventTypes.RoomAvatar,
+]
```
let's make this a set for efficient `in` performance:

```suggestion
STATE_EVENT_TYPES_TO_MARK_UNREAD = {
    EventTypes.PowerLevels,
    EventTypes.Topic,
    EventTypes.Name,
    EventTypes.RoomAvatar,
}
```

see: https://github.com/matrix-org/synapse/issues/7737

richvdh commented on synapse/storage/data_stores/main/events.py:
```diff
@@ -1614,3 +1626,87 @@ def f(txn, stream_ordering):
+        ):
+            return
+
+        body_exists = isinstance(content.get("body"), str)
```
this doesn't check for non-emptiness.

richvdh commented on synapse/storage/data_stores/main/events.py:
```diff
@@ -1614,3 +1626,87 @@ def f(txn, stream_ordering):
+        ):
+            return
+
+        body_exists = isinstance(content.get("body"), str)
+        is_state_event_to_mark_unread = (
+            event.is_state() and event.type in STATE_EVENT_TYPES_TO_MARK_UNREAD
+        )
+        is_encrypted_message = (
+            not event.is_state() and event.type == EventTypes.Encrypted
+        )
+
+        # We want to mark unread messages with a non-empty string body, some state events
+        # (power levels, room name, room topic, room avatar) and encrypted messages.
+        if not (body_exists or is_state_event_to_mark_unread or is_encrypted_message):
+            return
```
we can make this more efficient too. Factor out a separate `_should_mark_unread(event, context)` function (nb: can be a function rather than a method) and have it `return True`:

```
if context.rejected or event.internal_metadata.is_soft_failed():
    # exclude rejected and soft-failed events
    return False

# ... etc, other exclusions ...

if isinstance(body, str) and body != "":
    return True

if event.is_state() and event.type in STATE_EVENT_TYPES_TO_MARK_UNREAD:
    return True
```

richvdh commented on synapse/storage/data_stores/main/events.py:
```diff
@@ -1614,3 +1626,87 @@ def f(txn, stream_ordering):
+        )
+
+        # Mark the message as unread for every user currently in the room, except the
+        # sender of the event (because even if they haven't sent a read receipt for the
+        # event, it seems dumb to show it as unread to its sender).
```
this logic seems sane, but is probably worth making explicit in the MSC.

babolivier replied:
Done: https://github.com/matrix-org/matrix-doc/pull/2654/commits/928a8dff2efbd77064e3320d12b117a69f8169f4

richvdh commented on synapse/storage/data_stores/main/schema/delta/58/09unread_messages.sql:
```diff
@@ -0,0 +1,22 @@
+ * limitations under the License.
+ */
+
+CREATE TABLE IF NOT EXISTS unread_messages(
```
please could you add a comment which describes the purpose of this table?

richvdh commented on synapse/storage/data_stores/main/events_worker.py:
```diff
@@ -1362,6 +1362,62 @@ def get_next_event_to_expire_txn(txn):
+        # retrieved.
+        sql = """
+            SELECT COUNT(*) FROM unread_messages
+            WHERE user_id = ? AND room_id = ? AND stream_ordering > ?
```
I think a dedicated index on `(user_id, room_id, stream_ordering)` will help with this, especially for large accounts in lots of rooms.

richvdh commented on synapse/storage/data_stores/main/schema/delta/58/09unread_messages.sql:
```diff
@@ -0,0 +1,22 @@
+CREATE TABLE IF NOT EXISTS unread_messages(
+    user_id TEXT NOT NULL,  -- The user for which the message is unread.
+    stream_ordering BIGINT NOT NULL,  -- The position of the message in the event stream.
+    room_id TEXT NOT NULL,
+    event_id TEXT NOT NULL,  -- The ID of the message, we need it to handle redactions.
```
per #synapse-dev: I'm wondering  if we should have foreign key constraints here.

richvdh commented on synapse/storage/data_stores/main/events_worker.py:
```diff
@@ -1362,6 +1362,62 @@ def get_next_event_to_expire_txn(txn):
+            # before this user joined will be counted as well.
+            txn.execute(
+                """
+                SELECT stream_ordering FROM room_memberships
```
room_memberships contains all room membership events *ever*: if the user comes and goes a few times, there is no guarantee you'll get the right one.

I think you want `local_current_membership`, but check that it's not background-populated or something annoying.

babolivier replied:
Hmm right - I had updated `_maybe_insert_unread_event_txn` to use `current_state_events`, but apparently forgot that one.

richvdh commented on synapse/storage/data_stores/main/events_worker.py:
```diff
@@ -1362,6 +1362,62 @@ def get_next_event_to_expire_txn(txn):
             desc="get_next_event_to_expire", func=get_next_event_to_expire_txn
         )
 
+    async def get_unread_message_count_for_user(
```
I think this is going to need a cache for efficiency, and I think the cache will work best if you bring the logic for populating `last_read_event_id` inside the function.

of course, that means we'll have to worry about cache invalidation...

babolivier commented:
As pointed out in https://github.com/matrix-org/synapse/pull/7736/commits/420573a8ea744f42b3a111d406e254ea476a4b2d, this PR also changes the order replication streams are processed in, pasting the commit's description here for clarity:



> Receiving new data about e.g. events, read receipts, etc., is likely to
> result in some database actions. An example of that is that we send a
> push on new read receipts, which involves calculating a badge from
> unread counts. Because of this, we probably want to process cache
> updates/invalidations first so these actions are not using outdated
> cache data. In the previous example, we want to make sure we invalidate
> the cache on calculating the unread count for the user/room the read
> receipt is for before sending a push.

babolivier commented:
While implementing @richvdh's review, I've added caching to the `get_unread_message_count_for_user` function, and moved the call to `get_last_receipt_event_id_for_user` into it. This means we need to invalidate this cache every time we process a new read receipt. On the other hand, we systematically send a new push for each read receipt we process, so mobile clients know which messages have been read. This means we need to be sure that the cache invalidation for `get_unread_message_count_for_user` is processed before a new read receipt (otherwise we send pushes with wrong badge counts). This is impossible to enforce in a worker setup, because cache invalidations and read receipts are sent through replication on two different streams, and we have no way to ensure the order in which streams are processed.

To unwedge this, I think we should move the call to `get_last_receipt_event_id_for_user` back out of `get_unread_message_count_for_user` - for performance we could also merge this call with the one we use to calculate notification counts.

richvdh commented on synapse/storage/data_stores/main/events_worker.py:
```diff
@@ -1358,6 +1358,67 @@ def get_next_event_to_expire_txn(txn):
             desc="get_next_event_to_expire", func=get_next_event_to_expire_txn
         )
 
+    @cached(tree=True)
+    async def get_unread_message_count_for_user(
+        self, room_id: str, user_id: str,
+    ):
+        last_read_event_id = await self.get_last_receipt_event_id_for_user(
+            user_id=user_id, room_id=room_id, receipt_type="m.read",
+        )
```
you can specify that whenever the cache on `get_last_receipt_event_id_for_user` is invalidated, you want the cache on `get_unread_message_count_for_user` to be invalidated too:

```suggestion
    @cached(tree=True, cache_context=True)
    async def get_unread_message_count_for_user(
        self, room_id: str, user_id: str, cache_context: _CacheContext,
    ):
        last_read_event_id = await self.get_last_receipt_event_id_for_user(
            user_id=user_id, room_id=room_id, receipt_type="m.read",
            on_invalidate=cache_context.invalidate,
        )
```

I think that means that you can remove some of the explicit cache invalidation requests, and might solve your cache invalidation problem?

babolivier replied:
Oh I didn't know that! That might solve it yes, thanks!

richvdh commented on synapse/storage/data_stores/main/events.py:
```diff
@@ -53,6 +54,47 @@
+}
+
+
+def count_as_unread(event: EventBase, context: EventContext) -> bool:
```
this sounds like an order rather than a question.

```suggestion
def should_count_as_unread(event: EventBase, context: EventContext) -> bool:
```

richvdh commented on synapse/storage/data_stores/main/events.py:
```diff
@@ -176,6 +218,11 @@ def _persist_events_and_state_updates(
             )
             persist_event_counter.inc(len(events_and_contexts))
 
+            for event, _ in events_and_contexts:
+                self.store.get_unread_message_count_for_user.invalidate_many(
+                    (event.room_id,),
+                )
+
```
does this not also need to happen on worker instances? think you need to hook into `invalidate_caches_for_event`.

Also... given we already have a loop over `events_and_contexts` five lines down, any reason not to put this in there?

also also... can we skip it if `backfilled` ?

babolivier replied:
> Also... given we already have a loop over events_and_contexts five lines down, any reason not to put this in there?

I don't remember exactly, but I assume I just missed that loop.

> also also... can we skip it if backfilled ?

I don't think we should, because we might need to count backfilled events as unread. FWIW I'm usually trying to be consistent with how we handle push actions since it's a very similar feature, and I can see that we invalidate `get_unread_event_push_actions_by_room_for_user` regardless of `backfilled`.

richvdh commented on synapse/storage/data_stores/main/events.py:
```diff
@@ -1473,3 +1523,11 @@ def _update_backward_extremeties(self, txn, events):
         )
+
+    def _handle_redacted_unread_event_txn(self, txn: Connection, event: EventBase):
+        self.db.simple_update_txn(
```
there's no guarantee that you've received the redacted event before the redaction; does this behave correctly?

also: does it behave correctly if it turns out that the sender of the redaction didn't have permission to redact the event (noting that that cannot be enforced when the redaction is received because again: the redaction might be received over federation before the redacted event).

I am strongly minded to ignore redactions in the first pass. if it causes too many cases of "I got a ghost unread notification" then we can fix it later?

babolivier replied:
Yeah fair enough.

richvdh commented on synapse/storage/data_stores/main/events_worker.py:
```diff
@@ -1358,6 +1363,70 @@ def get_next_event_to_expire_txn(txn):
         )
 
+    @cached(tree=True, cache_context=True)
+    async def get_unread_message_count_for_user(
+        self, room_id: str, user_id: str, cache_context: _CacheContext,
+    ):
```
docstring please. In particular, what does it return, and what type?

richvdh commented on synapse/storage/data_stores/main/events_worker.py:
```diff
@@ -1358,6 +1363,70 @@ def get_next_event_to_expire_txn(txn):
+        user_id: str,
+        room_id: str,
+        last_read_event_id: Optional[str],
+    ):
```
can this have a return type annotation?

richvdh commented on synapse/storage/data_stores/main/events_worker.py:
```diff
@@ -1358,6 +1363,70 @@ def get_next_event_to_expire_txn(txn):
+
+    def _get_unread_message_count_for_user_txn(
+        self,
+        txn: LoggingTransaction,
```
Cursor is probably preferable here, unless you're particularly relying on the methods in LoggingTransaction

richvdh commented on synapse/storage/data_stores/main/events_worker.py:
```diff
@@ -1358,6 +1363,70 @@ def get_next_event_to_expire_txn(txn):
+                """,
+                (user_id, room_id),
+            )
+            row = txn.fetchone()
+            stream_ordering = row[0]
```
looks like this will explode horribly if the user turns out not to be a current member of the room. suggest failing gracefully somehow, or if you're really sure it shouldn't happen adding an assertion so that when it does the error message is better than obscure "ValueError".

richvdh commented on synapse/storage/data_stores/main/schema/delta/58/09unread_messages.sql:
```diff
@@ -0,0 +1,18 @@
+/* Copyright 2020 The Matrix.org Foundation C.I.C
```
this needs renaming I'm afraid. 09 has long been and gone.

babolivier commented on synapse/handlers/sync.py:
```diff
@@ -951,6 +952,17 @@ async def unread_notifs_for_room_id(
         # count is whatever it was last time.
         return None
 
+    async def unread_messages_for_room_id(
```
The reason for that was consistency with how we already calculate unread notif counts (in `unread_notifs_for_room_id`), though now that the read receipt retrieval has been moved to `get_unread_message_count_for_user` it probably makes more sense to remove that extra step yeah.

babolivier commented:
Oooh good call, I didn't think of that, thanks! Looks like both PostgreSQL and SQLite consider `NULL` as a false value, so simply removing the default Just Works™.

richvdh commented:
Note that this PR was later backed out in #8039, and re-done in #8059.
